### ISSUE TITLE
Critical bug in version 2.8.8 - unpublished news are listed - Contao 3.5.31

### ISSUE THREAD

**Issue by Total-Reality** [codefog/contao-news_categories#133]
Hey,

if a bot or a person which gets or type in a manipulated like www.example.com/news/category/contaoo/ then all news are listed, also the unpublished news. This is very critical and also a big duplicate content problem.
The real url is btw. www.example.com/news/category/contao/
The category "cantaoo" does'nt exist in the database...

I changed ModuleNewsList.php
you have a bug in the line 238 @ModuleNewsMenu.php

Before:
```
    public function generate()
    {
        if (TL_MODE == 'BE') {
            return parent::generate();
        }

        // Generate the list in related categories mode
        if ($this->news_relatedCategories) {
            return $this->generateRelated();
        }

        $GLOBALS['NEWS_FILTER_CATEGORIES'] = $this->news_filterCategories ? true : false;
        $GLOBALS['NEWS_FILTER_DEFAULT']    = deserialize($this->news_filterDefault, true);
        $GLOBALS['NEWS_FILTER_PRESERVE']   = $this->news_filterPreserve;

        $buffer = parent::generate();

        // Cleanup the $GLOBALS array (see #57)
        unset($GLOBALS['NEWS_FILTER_CATEGORIES'], $GLOBALS['NEWS_FILTER_DEFAULT'], $GLOBALS['NEWS_FILTER_PRESERVE']);

        return $buffer;
    }
```

After:

```
    public function generate()
    {
        if (TL_MODE == 'BE') {
            return parent::generate();
        }

        // Generate the list in related categories mode
        if ($this->news_relatedCategories) {
            return $this->generateRelated();
        }

        if(\Input::get('category'))
        {
          // Find out if category exists
          $intId = \NewsCategoryModel::findPublishedByIdOrAlias(\Input::get('category'))->id;

          if($intId)
          {
            $GLOBALS['NEWS_FILTER_CATEGORIES'] = $this->news_filterCategories ? true : false;
            $GLOBALS['NEWS_FILTER_DEFAULT']    = deserialize($this->news_filterDefault, true);
            $GLOBALS['NEWS_FILTER_PRESERVE']   = $this->news_filterPreserve;
          }
          // Redirect if category doesn't exist
          else
          {
            global $objPage;
            $strUrl = \Controller::generateFrontendUrl($objPage->row());
            $this->redirect($strUrl);
            exit();
          }
        }

        $buffer = parent::generate();

        // Cleanup the $GLOBALS array (see #57)
        unset($GLOBALS['NEWS_FILTER_CATEGORIES'], $GLOBALS['NEWS_FILTER_DEFAULT'], $GLOBALS['NEWS_FILTER_PRESERVE']);

        return $buffer;
    }
```

It would be great if you can fix this as far as possible.

Best regards,
Chris

**Comment by Total-Reality** [comment]
Thanks, but it is not fully fixed. You should redirect, because of duplicate content... 
www.example.com/blog/kategorie/blaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/ is very senseless at all...

**Comment by qzminski** [comment]
Yeah, you're right. Fixed in 1ba059a.

**Comment by Total-Reality** [comment]
Thank you :)

One note left:
A 404 status is contao standard, that's right... however it is not so much better than a duplicate or empty page. The problem is that Google don't like 404 errors. 
It is very easy to create a redirect in a contao module like ModuleNewsList.php to the parent page if the category does not exist. But it is very difficult to redirect it with htaccess. Because redirects via htaccess mean that you have to be careful and should not redirect existing pages. I don't know how to solve this problem... although I programmed several difficult things in htaccess.
Or have you an idea how to you that?

It is the same thing like news, I don't know why contao gives out a 404-page in the news reader. That makes no sense at all.

**Comment by fritzmg** [comment]
This is not true. You should definitely __not__ redirect an invalid URL. An invalid URL needs to respond with a 404 status.

**Comment by Total-Reality** [comment]
Nope... I think you have not much knowledge in SEO? But no problem... many people who are programming for Contao have the same problem.
If you would know webmaster tools, you should not say sth. like this. What is the advantage for the user seeing a 404 page? 

Unbelievable...

**Comment by Total-Reality** [comment]
Additionally this is no classic situation like a missing page. It is a mistake made by the programmer or laziness of him, because he allows the user to link to a not existing category page. It is a "auto_item" Bug. 
It is the same that sb. allows ?page=100000 - although there are existing 10 pages only.

A nice implementation can be found here:
http://www.spiegel.de/wirtschaft/unternehmen/anton-schlecker-eine-milliardfhgfgfgffge-euro-schulden-bleiffgfgfgfgben-bestehen-a-1190083.html

Redirects to...
http://www.spiegel.de/wirtschaft/unternehmen/anton-schlecker-eine-milliarde-euro-schulden-bleiben-bestehen-a-1190083.html

**Comment by fritzmg** [comment]
> Nope... I think you have not much knowledge in SEO? But no problem... many people who are programming for Contao have the same problem.

Why the hostility?

> If you would know webmaster tools, you should not say sth. like this. What is the advantage for the user seeing a 404 page?

The user never sees this 404 page because invalid links typically aren't generated in the front end and thus indexed in the first place. If an invalid URL lands in the index of a search engine for some reason or another, the search engine needs to be told that this URL is invalid. You do that with a 404 status.

A 301 redirect on the other hand tells the search engine that the actual content of the (actually invalid) URL is now to be found under the new URL. Which of course does not make sense for any invalid URL.

The Google Search Console tells you whenever any indexed URL responds with a 404 page so that you can check the status of that URL and take appropriate measures. I.e. fix the problem (if there is one), redirect to a new URL if the content is now found under a different URL or keep the 404 status and may be even remove the URL manually from the index.

> not existing category page

Hence a 404 status code should be generated. Such URL should neither be index nor redirect to somewhere else.

> A nice implementation can be found here:
> http://www.spiegel.de/wirtschaft/unternehmen/anton-schlecker-eine-milliardfhgfgfgffge-euro-schulden-bleiffgfgfgfgben-bestehen-a-1190083.html

Just because other do it does not mean it's correct or good in any way.

**Comment by Total-Reality** [comment]
Because you said this is not true. You don't define what is true or not...

You have good arguments, however this is not enough. Because your logic is not user friendly + also not search engine friendly + also not webmaster friendly. I don't want to have hundreds of 404 errors in my Systemlog, so I need to overwrite this class/module. 

Very good work...

**Comment by fritzmg** [comment]
My "This is not true." was regarding this comment:

> The problem is that Google don't like 404 errors.

This is simply not true. Or rather a misconception.

> Because your logic is not user friendly

It would __not__ be user friendly to redirect the user to an URL which has nothing to do with the original URL. If the original URL simply never existed, because it is simply invalid, or if the content does not exist anymore, a redirect would not be helpful for the user.

> also not search engine friendly

It __is__ search engine friendly, because if a non-sensical URL is indexed by the search engine for some reason, the search engine needs to know that this URL does not contain any content and thus should not be in the index anymore.

> also not webmaster friendly

Why would it be webmaster friendly if the news_categories module would automatically redirect to some page? It's more helpful to you that a non-existing URL responds with a 404 status.

> I don't want to have hundreds of 404 errors in my Systemlog

Why would you need to do that? There won't be any regular access to such non-sensical URLs since they aren't generated in the front end.

Btw. Contao 4 does not log 404 responses anymore.

**Comment by qzminski** [comment]
I'm with @fritzmg here. If the category does not exist then it means that the resource is not found which is represented by 404 status.
